src/components/queryBuilder/EditorTypeSwitcher.tsx

```diff
@@ -0,0 +1,108 @@
+import React, { useState } from 'react';
+import { SelectableValue } from '@grafana/data';
+import { RadioButtonGroup, ConfirmModal, InlineFormLabel } from '@grafana/ui';
+import { getQueryOptionsFromSql, getSQLFromQueryOptions } from '../queryBuilder/utils';
+import labels from 'labels';
+import { EditorType, CHQuery, defaultCHBuilderQuery, CHSqlQuery } from 'types/sql';
+import { QueryBuilderOptions } from 'types/queryBuilder';
+import isString from 'lodash/isString';
+
+interface CHEditorTypeSwitcherProps {
+  query: CHQuery;
+  onChange: (query: CHQuery) => void;
+  onRunQuery: () => void;
+}
+
+/**
+ * Component for switching between the SQL and Query Builder editors.
+ */
+export const EditorTypeSwitcher = (props: CHEditorTypeSwitcherProps) => {
+  const { query, onChange } = props;
+  const { label, tooltip, switcher, cannotConvert } = labels.components.EditorTypeSwitcher;
+  let editorType: EditorType =
+    query.editorType ||
+    ((query as CHSqlQuery).rawSql && !(query as CHQuery).editorType ? EditorType.SQL : EditorType.Builder);
+  const [editor, setEditor] = useState<EditorType>(editorType);
+  const [confirmModalState, setConfirmModalState] = useState<boolean>(false);
+  const [cannotConvertModalState, setCannotConvertModalState] = useState<boolean>(false);
+  const options: Array<SelectableValue<EditorType>> = [
+    { label: labels.types.EditorType.sql, value: EditorType.SQL },
+    { label: labels.types.EditorType.builder, value: EditorType.Builder },
+  ];
+  const [errorMessage, setErrorMessage] = useState<string>('');
+  const onEditorTypeChange = (editorType: EditorType, confirm = false) => {
+    if (query.editorType === EditorType.SQL && editorType === EditorType.Builder && !confirm) {
+      const queryOptionsFromSql = getQueryOptionsFromSql(query.rawSql);
+      if (isString(queryOptionsFromSql)) {
+        setCannotConvertModalState(true);
+        setErrorMessage(queryOptionsFromSql);
+      } else {
+        setConfirmModalState(true);
+      }
+    } else {
+      setEditor(editorType);
+      let builderOptions: QueryBuilderOptions;
+      switch (query.editorType) {
+        case EditorType.Builder:
+          builderOptions = query.builderOptions;
+          break;
+        case EditorType.SQL:
+          builderOptions =
+            (getQueryOptionsFromSql(query.rawSql) as QueryBuilderOptions) || defaultCHBuilderQuery.builderOptions;
+          break;
+        default:
+          builderOptions = defaultCHBuilderQuery.builderOptions;
+          break;
+      }
+      if (editorType === EditorType.SQL) {
```

Comment:
If we always have to check the type within the callback, maybe we should just use 2 different callbacks?